### DISCUSSION TITLE
Fork flutter_markdown

### DISCUSSION THREAD

**Discussion by micahmo** [thunder-app/thunder#860] [General]
I would like to discuss the possibility of eventually forking [flutter_markdown](https://pub.dev/packages/flutter_markdown). I believe this might be necessary to achieve a few of the features that we want and that have been requested. Here are the ones I'm thinking of, and there may be more.

1. Support for spoiler syntax. I did some investigation [here](https://github.com/thunder-app/thunder/issues/304#issuecomment-1660486274) and found that, despite the cool ability to supply markdown extensions, if the HTML output contains an unrecognized tab (in this case `details`), then it won't render properly. Support for that tag will need to be added upstream.
2. I would like to add an InkWell to links.
3. I would like to add a long-press menu on links (e.g., with the ability to copy, share, see the target URL, etc.).
   * #709, #309

**Comment by hjiangsu** [comment]
> Support for spoiler syntax. I did some investigation https://github.com/thunder-app/thunder/issues/304#issuecomment-1660486274 and found that, despite the cool ability to supply markdown extensions, if the HTML output contains an unrecognized tab (in this case details), then it won't render properly. Support for that tag will need to be added upstream.

Hmm, I wouldn't be opposed to this idea. However, I'd like to try it myself with the extension and see if there's perhaps an alternative to forking the package. If we do end up choosing to fork the package, we should keep in mind to make it as easy to possible to receive upstream changes in the cases where there's an update!

> I would like to add an InkWell to links.
> I would like to add a long-press menu on links (e.g., with the ability to copy, share, see the target URL, etc.).

I haven't played around with it yet but I wonder if we can achieve these features by using `MarkdownElementBuilder` provided by `lutter_markdown`. [Here's the API reference for it.](https://pub.dev/documentation/flutter_markdown/latest/flutter_markdown/MarkdownElementBuilder-class.html)

**Comment by micahmo** [comment]
> I'd like to try it myself with the extension

Please do! I would be glad to be wrong, as I know maintaining another fork would be extra work.

Keep in mind that the extensions themselves are not the problem. They're very powerful and let you basically write your own parser to detect a sequence and replace it with a different output. I was able to write an extension for spoilers which outputs perfect HTML. When I saved that HTML to a file and opened in the browser, it worked perfectly. 

The problem is actually in the HTML renderer. I'm assuming we'd want to use the [`details` HTML tag](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_details) for collapsible content (which is what the web UI uses). You can see in the [list of supported tags](https://github.com/flutter/packages/blob/main/packages/flutter_markdown/lib/src/style_sheet.dart), `details` isn't one of them. So if we made a change, I think that's where it would be! See [my original comment](https://github.com/thunder-app/thunder/issues/304#issuecomment-1660486274) for more info.

> I wonder if we can achieve these features by using `MarkdownElementBuilder`

That looks very promising!!

Thanks for the good discussion!

**Comment by micahmo** [comment]
Thanks to your research, I think we have determined that we do not need to fork [flutter_markdown](https://pub.dev/packages/flutter_markdown). For spoilers, see #867. For link improvements, see #869.
